### PR TITLE
Create solutions from existing items (part of #151)

### PR DESCRIPTION
**Pull request by flange-ipb**

### CODE REVIEW

File: ui/src/main/java/de/ipb_halle/lbac/items/Item.java [hunk 1]
```diff
@@ -377,4 +391,14 @@ public void setLabel(String label) {
         this.label = label;
     }
 
+    /**
+     * @return new quantity with the {@code amount} as value and the
+     * {@code unit} as unit or null if one of those is null
+     */
+    public Quantity getAmountAsQuantity() {
```
Comment (FabianMauz) [inline ui/src/main/java/de/ipb_halle/lbac/items/Item.java:398]: Use method of Factory?

File: ui/src/main/java/de/ipb_halle/lbac/items/bean/createsolution/consumepartofitem/ConsumePartOfItemStep1Controller.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            targetConcentrationUnit = Unit.getUnit("mM");
+        } else {
+            availableConcentrationUnits.addAll(massConcentrations);
+            targetConcentrationUnit = Unit.getUnit("g/l");
+        }
+
+        targetVolumeUnit = Unit.getUnit("ml");
+
+        Quantity massFromItem = massFromParentItem();
+        if (massFromItem != null) {
```
Comment (FabianMauz) [inline ui/src/main/java/de/ipb_halle/lbac/items/bean/createsolution/consumepartofitem/ConsumePartOfItemStep1Controller.java:91]: is always not null

File: ui/src/main/java/de/ipb_halle/lbac/items/bean/createsolution/consumepartofitem/ConsumePartOfItemStep1Controller.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    }
+
+    /*
+     * Getters with logic
+     */
+    public boolean isTargetMassGreaterThanItemMass() {
+        Quantity massFromItem = massFromParentItem();
+        Quantity targetMassAsQuantity = new Quantity(targetMass, targetMassUnit);
+
+        if (targetMassAsQuantity.isGreaterThan(massFromItem)) {
```
Comment (FabianMauz) [inline ui/src/main/java/de/ipb_halle/lbac/items/bean/createsolution/consumepartofitem/ConsumePartOfItemStep1Controller.java:148]: shorten, only return needed ;)

File: ui/src/main/java/de/ipb_halle/lbac/items/bean/createsolution/consumepartofitem/ConsumePartOfItemStep2Controller.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    public Double getTargetMassMinusMargin() {
+        Double targetMass = step1Controller.getTargetMass();
+        if (targetMass == null) {
+            return null;
+        }
+        return targetMass * (1.0 - (errorMargin / 100.0));
+    }
+
+    public boolean isWeighGreaterThanItemMass() {
```
Comment (FabianMauz) [inline ui/src/main/java/de/ipb_halle/lbac/items/bean/createsolution/consumepartofitem/ConsumePartOfItemStep2Controller.java:85]: typo -> weight

File: ui/src/main/java/de/ipb_halle/lbac/items/bean/createsolution/consumepartofitem/ConsumePartOfItemStep2Controller.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        Quantity massFromItem = parentItem.getAmountAsQuantity();
+        Quantity weighAsQuantity = new Quantity(weigh, weighUnit);
+
+        if (weighAsQuantity.isGreaterThan(massFromItem)) {
```
Comment (FabianMauz) [inline ui/src/main/java/de/ipb_halle/lbac/items/bean/createsolution/consumepartofitem/ConsumePartOfItemStep2Controller.java:89]: shorten

File: ui/src/main/java/de/ipb_halle/lbac/items/bean/createsolution/consumepartofitem/ConsumePartOfItemStep2Controller.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    public Quantity getWeighAsQuantity() {
```
Comment (FabianMauz) [inline ui/src/main/java/de/ipb_halle/lbac/items/bean/createsolution/consumepartofitem/ConsumePartOfItemStep2Controller.java:96]: typo

File: ui/web/WEB-INF/templates/item/createSolution/consumePartOfItemStep4.xhtml [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+                <b:row>
+                    <b:column span="1">
+                        <b:button
+                            value="#{msgs.wizard_back}"
+                            onclick="PF('consumePartOfItemWizard').back();" />
+                    </b:column>
+                    <b:column
+                        span="1"
+                        offset="9">
+                        <b:commandButton
```
Comment (flange-ipb) [inline ui/web/WEB-INF/templates/item/createSolution/consumePartOfItemStep4.xhtml:41]: Bug: The actionSave() method does not validate that the container size is greater than the weight like it's done when jumping from step4 to step5.

File: ui/src/main/java/de/ipb_halle/lbac/util/calculation/MolarConcentrationCalculations.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+     * @return mass m in the given unit
+     */
+    public static Quantity calculateMass(Quantity molarConcentration, Quantity molarMass, Quantity volume,
+            Unit massUnit) {
+        massUnit.require(MASS);
+        return calculateMass(molarConcentration, molarMass, volume).to(massUnit);
+    }
+
+    /**
+     * V = m / (c * M)
```
Comment (FabianMauz) [inline ui/src/main/java/de/ipb_halle/lbac/util/calculation/MolarConcentrationCalculations.java:100]: change formula to match code

File: ui/src/main/java/de/ipb_halle/lbac/util/units/Quantity.java [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+    }
+
+    /**
+     * Factory method (null-safe)
+     * 
+     * @param value
+     * @param unit
+     * @return new quantity or null if value or unit are null
+     */
+    public static Quantity create(Double value, Unit unit) {
```
Comment (FabianMauz) [inline ui/src/main/java/de/ipb_halle/lbac/util/units/Quantity.java:48]: why factory method and public constructor?

File: ui/src/main/resources/de/ipb_halle/lbac/i18n/messages_de.properties [hunk 10]
```diff
@@ -385,6 +385,30 @@ itemAgent_label_search=Suchen
...
+itemCreateSolution_consumePartOfItem_dispensedVolume=Dispensiertes Volumen 
+itemCreateSolution_consumePartOfItem_errorMargin=Fehlerspanne
+itemCreateSolution_consumePartOfItem_finalConcentration=Finale Konzentration
+itemCreateSolution_consumePartOfItem_solvent=Lösungsmittel
+itemCreateSolution_consumePartOfItem_targetConcentration=Zielkonzentration
+itemCreateSolution_consumePartOfItem_targetMass=Zielmasse
+itemCreateSolution_consumePartOfItem_targetVolume=Zielvolumen
+itemCreateSolution_consumePartOfItem_unit=Einheit
+itemCreateSolution_consumePartOfItem_volumeToDispense=Dosiervolumen
+itemCreateSolution_consumePartOfItem_weigh=Einwaage
```
Comment (FabianMauz) [inline ui/src/main/resources/de/ipb_halle/lbac/i18n/messages_de.properties:404]: typo : weight

File: ui/src/main/resources/de/ipb_halle/lbac/i18n/messages_en.properties [hunk 11]
```diff
@@ -385,6 +385,30 @@ itemAgent_label_search=Search
 itemAgent_table_action=Action
 itemAgent_table_insert=Insert
 itemAgent_table_quantity=Available quantity
+itemCreateSolution_consumePartOfItem_step1Title=Target concentration and volume
+itemCreateSolution_consumePartOfItem_step2Title=Weigh
```
Comment (FabianMauz) [inline ui/src/main/resources/de/ipb_halle/lbac/i18n/messages_en.properties:389]: typo

File: ui/src/test/java/de/ipb_halle/lbac/items/mocks/LabelServiceMock.java [hunk 12]
```diff
@@ -17,13 +17,13 @@
...
 
-import de.ipb_halle.lbac.label.LabelService;
+import de.ipb_halle.lbac.items.service.ItemLabelService;
 
 /**
  *
  * @author fmauz
  */
-public class LabelServiceMock extends LabelService {
+public class LabelServiceMock extends ItemLabelService {
```
Comment (flange-ipb) [inline ui/src/test/java/de/ipb_halle/lbac/items/mocks/LabelServiceMock.java:26]: rename class to ItemLabelServiceMock

**Pull request merged**